### PR TITLE
Metadata V15: Enrich extrinsic type info for decoding

### LINKED ISSUE #12929: V14 metadata doesn't provide enough information to decode extrinsic
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Description of bug

To properly decode extrinsic one needs to know

1. The type of `Address`
2. The type of `Signature`
3. The type of `Call`.

While all the above types are included in the `.lookup.types`, there is no way to know which one is which.

At Subsquid we look [at paths and type parameters ](https://github.com/subsquid/squid-sdk/blob/ab1cae1eb6c9c4a34760f4fea09dff45c2f4065b/substrate/substrate-metadata/src/chainDescription.ts#L170).

Polkadot.js uses its own set of hardcoded assumptions none of which is guaranteed to hold.

This is practical issue which hardens the maintenance of our stack.

### Steps to reproduce

_No response_

### PR DESCRIPTION
**Pull request by lexnv**
### Extrinsic Metadata fields

Metadata is enriched to provide users with additional types for decoding extrinsics:
- `Address` type 
- `Call` type 
- `Signature` type 
- `Extra` type 

Although already available on the extrinsic's type generic parameters, having the fields
readily available provides enough information for users to decode extrinsics without
making assumptions about the metadata shape (e.g., this eliminates the need of subxt
to parse those fields manually [here](https://github.com/paritytech/subxt/blob/3f16bb8d521031c7e13749d41711f54f490bb84c/subxt/src/blocks/extrinsic_types.rs#L483-L526), feature also requested by the linked issue). 

Note: the `Extra` type similar to others can be used to decode the extrinsics, as well as to
skip decoding over those fields (see [subxt's decoding](https://github.com/paritytech/subxt/blob/3f16bb8d521031c7e13749d41711f54f490bb84c/subxt/src/blocks/extrinsic_types.rs#L259-L265)).

### Incomplete UncheckedExtrisnic type

It may be possible for certain substrate-based chains, or even substrate tests, to construct a
runtime with different types, or missing types. For those edge cases, the type reported back in
the metadata is the type of the tuple `meta_type::<()>()`.

Depends on frame-metadata release.
Closes https://github.com/paritytech/substrate/issues/12929.

### Subxt Testing

- Fetched the unstable metadata
- Type IDs match the extrinsic type params

```bash
# This is the extrinsic type with enriched info
"extrinsic": {
    "ty": 786,
    "version": 4,
    "address_ty": 161,
    "call_ty": 135,
    "signature_ty": 348,
    "extra_ty": 787,

# This is the extrinsic.ty which has the type params exactly the
# same as what is reported in the extrinsic for address, call, signature, extra
    "id": 786,
    "type": {
      "path": [
        "sp_runtime",
        "generic",
        "unchecked_extrinsic",
        "UncheckedExtrinsic"
      ],
      "params": [
        {
          "name": "Address",
          "type": 161
        },
        {
          "name": "Call",
          "type": 135
        },
        {
          "name": "Signature",
          "type": 348
        },
        {
          "name": "Extra",
          "type": 787
        }
      ],
```

// @paritytech/subxt-team @jsdw @bkchr

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2819994

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2840377

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2840515

File: frame/system/src/offchain.rs [hunk 1]
```diff
@@ -86,7 +86,11 @@ where
 	/// Submit transaction onchain by providing the call and an optional signature
 	pub fn submit_transaction(
 		call: <T as SendTransactionTypes<LocalCall>>::OverarchingCall,
-		signature: Option<<T::Extrinsic as ExtrinsicT>::SignaturePayload>,
+		signature: Option<(
```
Comment (kianenigma) [inline frame/system/src/offchain.rs:89]: I don't get why we have to break down `SignaturePayload`? can't this have `TypeInfo` itself? 

And if it is being broken down and not used, I rather actually delete `SignaturePayload` if it is not being used anymore.

File: primitives/runtime/src/generic/unchecked_extrinsic.rs [hunk 2]
```diff
@@ -106,13 +106,20 @@ impl<Address, Call, Signature, Extra: SignedExtension> Extrinsic
 {
 	type Call = Call;
 
-	type SignaturePayload = (Address, Signature, Extra);
```
Comment (kianenigma) [inline primitives/runtime/src/generic/unchecked_extrinsic.rs:109]: ahh, it is being done already 🙈

Comment (kianenigma) [reply primitives/runtime/src/generic/unchecked_extrinsic.rs:109]: In this case, a `type SignaturePayloadOf<E> = (<E as Extrinsic>::X, <E as Extrinsic>::Y, <E as Extrinsic>::Z)` will help prevent potential error where the order of these 3 might be mistaken.

**Review by kianenigma — COMMENTED** [review]
Looks overall good, just one suggestion about using a type alias.

**Comment by jsdw** [comment]
Does `extrinsic.ty` have any use I wonder with this change? In Subxt at least, the only reason to ever inspect the type was to find out the type IDs of the Signature/Extra/etc bits, which are now handily provided by this change :)

File: primitives/runtime/src/traits.rs [hunk 3]
```diff
@@ -961,12 +961,20 @@ pub trait Extrinsic: Sized {
 	/// The function call.
 	type Call;
 
-	/// The payload we carry for signed extrinsics.
+	/// The type of the address that signed the extrinsic.
 	///
-	/// Usually it will contain a `Signature` and
-	/// may include some additional data that are specific to signed
-	/// extrinsics.
-	type SignaturePayload;
```
Comment (bkchr) [inline primitives/runtime/src/traits.rs:969]: Instead of breaking most of the code we could just do
```
type SignaturePayload: SignaturePayload;
```

```
trait SignaturePayload {
      type Address: TypeInfo;
      type Signature: TypeInfo;
      type Extra: TypeInfo;
}
```

Comment (lexnv) [reply primitives/runtime/src/traits.rs:969]: Yep, I like this approach more! Thanks for the feedback!

Comment (lexnv) [reply primitives/runtime/src/traits.rs:969]: @bkchr What's your opinion on having the `extrinsic.ty` in the metadata after this change (the message James mentioned above)?

We should have enough type info by relying on:
- address_ty
- call_ty
- signature_ty
- extra_ty

The `extrinsic.ty` has a manual implementation of `TypeInfo` that adds the above ty as generic types, but since we explicitly expose that info with this PR, I believe we should remove the `extrinsic.ty`.

That is unless we'll break other things

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2910974

File: primitives/runtime/src/traits.rs (hunk 3 above)
Comment (bkchr) [reply primitives/runtime/src/traits.rs:972]: I have no idea if someone is using this. However, frankly speaking by only having the types you listed you don't know the encoding of the `Extrinsic` type. While currently it may also already not being correct as it probably doesn't include the `Compact<u32>` length in from of the extrinsinc encoding?

Comment (lexnv) [reply primitives/runtime/src/traits.rs:972]: Good idea! I'll add the `TypeDefCompact<u32>` type to the extrinsic metadata to provide even better decoding abilities to end users. Considering this, I'll remove the `extrinsic_ty` from the metadata, unless encountering some edge-cases from UX teams.

The type only contained the `address`,  `call`, `signature` and `extra` on its type info definition, and users should be able to obtain those details (and more) directly.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2915798

File: primitives/runtime/src/traits.rs (hunk 3 above)
Comment (jsdw) [reply primitives/runtime/src/traits.rs:972]: iirc the extrinsic type was quite useless anyway before; it mostly just said "hey we want to encode the extrinsic to some vec of bytes". It doesn't provide any useful information on how those bytes need to be formed etc.

There are already a bunch of things we must just assume already about encoding extrinsics, ie the order of the signature stuff, how the not-really-scale Option for the signature is encoded (ie it has a version number `&&`'d to it) etc. I've always assumed this format information to be encoded in the tx version (4 atm); if a transaction has this version, it must be encoded in this way. 

The metadata, in my mind, only needs to provide the information that might vary from chain to chain (eg the actual additional params etc). This tx version can encode the actual format that this stuff is turned into an extrinsic.

All of this to say, what would adding a  `TypeDefCompact<u32>` to the metadata help with? I'd assume it to be part of the V4 extrinsic format I guess, but perhaps I'm wrong on this :)

Comment (bkchr) [reply primitives/runtime/src/traits.rs:972]: > There are already a bunch of things we must just assume already about encoding extrinsics, ie the order of the signature stuff, how the not-really-scale Option for the signature is encoded (ie it has a version number `&&`'d to it) etc. I've always assumed this format information to be encoded in the tx version (4 atm); if a transaction has this version, it must be encoded in this way.

Good point. 

> The metadata, in my mind, only needs to provide the information that might vary from chain to chain (eg the actual additional params etc). This tx version can encode the actual format that this stuff is turned into an extrinsic.

Just because every body is currently using our `UncheckedExtrinsic` doesn't mean that this will always be the case. However, it makes sense that this is a type that we can not depict in as type metadata.

File: frame/support/procedural/src/construct_runtime/expand/metadata.rs [hunk 4]
```diff
@@ -95,11 +95,31 @@ pub fn expand_runtime_metadata(
...
+				let call_ty = #scrate::scale_info::meta_type::<
+					<#extrinsic as #scrate::sp_runtime::traits::Extrinsic>::Call
+					>();
+				let signature_ty = #scrate::scale_info::meta_type::<
+						<<#extrinsic as #scrate::sp_runtime::traits::Extrinsic>::SignaturePayload as #scrate::sp_runtime::traits::SignaturePayload>::Signature
+					>();
+				let extra_ty = #scrate::scale_info::meta_type::<
+						<<#extrinsic as #scrate::sp_runtime::traits::Extrinsic>::SignaturePayload as #scrate::sp_runtime::traits::SignaturePayload>::SignatureExtra
+					>();
+				let len_ty = #scrate::scale_info::meta_type::<#scrate::codec::Compact<u32>>();
```
Comment (bkchr) [inline frame/support/procedural/src/construct_runtime/expand/metadata.rs:111]: Based on the comment of @jsdw we should not add this.

File: primitives/runtime/src/traits.rs [hunk 5]
```diff
@@ -988,6 +988,30 @@ pub trait Extrinsic: Sized {
 	}
 }
 
+/// Something that acts like a `SignaturePayload` of an `Extrinsic.
```
Comment (bkchr) [inline primitives/runtime/src/traits.rs:991]:
```suggestion
/// Something that acts like a [`SignaturePayload`](Extrinsic::SignaturePayload) of an [`Extrinsic`].
```

**Review by bkchr — APPROVED** [review]

File: frame/support/procedural/src/construct_runtime/expand/metadata.rs (hunk 4 above)
Comment (lexnv) [reply frame/support/procedural/src/construct_runtime/expand/metadata.rs]: I've removed the `len_ty` form the V15 conversion, thanks!

**Comment by lexnv** [comment]
bot rebase

**Review by jsdw — APPROVED** [review]
Looks great; well done!

Remember to release frame-metadata etc before merging :)

**Review by sam0x17 — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Comment by lexnv** [comment]
Polkadot companion is expected to fail since the latest companion conflicts, but will be solved by https://github.com/paritytech/polkadot/pull/7401#issuecomment-1611128959

**Comment by lexnv** [comment]
bot merge
